kernel/ps2/src/lib.rs

```diff
@@ -1,484 +1,704 @@
-        if let Err(_e) = command_to_mouse(value) {
-            Err("set mouse sampling rate failled, please try again")
+    // implementation from https://wiki.osdev.org/PS/2_Mouse
+    fn to_2s_complement(bit1to8: u8, bit9: bool) -> i16 {
+        // to fit a 9th bit inside; we can't just convert to i16, because it would turn e.g. 255 into -1
+        let unsigned = bit1to8 as u16; // 1111_1111 as u16 = 0000_0000_1111_1111
+
+        // now convert into i16, which always gives us a positive number
+        let signed = unsigned as i16; // 0000_0000_1111_1111 as i16 = 0000_0000_1111_1111
+        
+        if bit9 {
+            // value is negative, produce the two's complement, correctly sign extended no matter its size
+            signed - 0b1_0000_0000 // 0000_0000_1111_1111 - 1_0000_0000 = -1
         } else {
-            Ok(())
+            signed
         }
     }
+
+    //currently limited to mouse id 4
+    /// often called `z_movement`, renamed to disambiguate
+    pub fn scroll_movement(&self) -> i8 {
+        self.z_movement() as i8 //TODO: see osdev wiki, might be wrong
+    }
 }
 
-/// set the mouse ID (3 or 4 ) by magic sequence
-/// 3 means that the mouse has scroll
-/// 4 means that the mouse has scroll, fourth and fifth buttons
-pub fn set_mouse_id(id: u8) -> Result<(), &'static str> {
-    // stop the mouse's streaming before trying to set the ID
-    // if fail to stop the streaming, return error
-    if let Err(_e) = mouse_packet_streaming(false) {
-        warn!("fail to stop streaming, before trying to read mouse id");
-        return Err("fail to set the mouse id!!!");
-    } else {
-        // set the id to 3
-        if id == 3 {
-            if let Err(_e) = set_sampling_rate(200) {
-                warn!("fail to stop streaming, before trying to read mouse id");
-                return Err("fail to set the mouse id!!!");
-            }
-            if let Err(_e) = set_sampling_rate(100) {
-                warn!("fail to stop streaming, before trying to read mouse id");
-                return Err("fail to set the mouse id!!!");
-            }
-            if let Err(_e) = set_sampling_rate(80) {
-                warn!("fail to stop streaming, before trying to read mouse id");
-                return Err("fail to set the mouse id!!!");
+/// read mouse data packet; will work for mouse with ID 4 and probably 3
+pub fn read_mouse_packet() -> MousePacketBits4 {
+    MousePacketBits4::from_bytes([
+        read_data(),
+        read_data(),
+        read_data(),
+        read_data()
+    ])
+}
+
+#[derive(Clone)]
+pub enum SampleRate {
+    _10 = 10,
+    _20 = 20,
+    _40 = 40,
+    _60 = 60,
+    _80 = 80,
+    _100 = 100,
+    _200 = 200
+}
+
+/// set ps2 mouse's sampling rate
+fn set_mouse_sampling_rate(value: SampleRate) -> Result<(), &'static str> {
+    command_to_mouse(HostToMouseCommandOrData::MouseCommand(SampleRate))
+        .and(command_to_mouse(HostToMouseCommandOrData::SampleRate(value)))
```

Comment:
i think this needs to be `and_then`, not `and`, because I assume you only want to invoke the second command if the first command succeeds. 

Currently this will result in both commands being run regardless of the first one's success.